rbx_types/src/terrain.rs

```diff
@@ -0,0 +1,548 @@
+/// Relevant for usage with a `Terrain` object. Inside a grid of 524,288^3 chunks per world.
+#[repr(transparent)]
+#[derive(Debug, PartialEq, Eq, Default, Clone, Copy, PartialOrd, Ord, Hash)]
+#[cfg_attr(
+    feature = "serde",
+    derive(serde::Serialize, serde::Deserialize),
+    serde(transparent)
+)]
+pub struct ChunkCoordinates(TerrainVec);
+
+impl ChunkCoordinates {
+    /// Constructs a new `ChunkCoordinates` object.
+    #[inline]
+    pub fn new(x: i32, y: i32, z: i32) -> Self {
+        // Terrain isn't accepted by the engine with voxels past 2^23.
+        const CHUNK_MAX: i32 = 2i32.pow(23) / CHUNK_SIZE;
+        let (x, y, z) = (
+            x.clamp(-CHUNK_MAX, CHUNK_MAX),
+            y.clamp(-CHUNK_MAX, CHUNK_MAX),
+            z.clamp(-CHUNK_MAX, CHUNK_MAX),
+        );
+        Self(TerrainVec::new(x, y, z))
+    }
+
+    /// Constructs a new `ChunkCoordinates` object from a Vector3.
+    #[inline]
+    pub fn from_vec3(i: Vector3) -> Self {
+        Self(TerrainVec::from_vec3(i))
+    }
+}
+
+#[repr(u8)]
+#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
+#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
+pub enum Material {
+    #[default]
+    Air = 0x00,
+    Water = 0x01,
+    Grass = 0x02,
+    Slate = 0x03,
+    Concrete = 0x04,
+    Brick = 0x05,
+    Sand = 0x06,
+    WoodPlanks = 0x07,
+    Rock = 0x08,
+    Glacier = 0x09,
+    Snow = 0x0A,
+    Sandstone = 0x0B,
+    Mud = 0x0C,
+    Basalt = 0x0D,
+    Ground = 0x0E,
+    CrackedLava = 0x0F,
+    Asphalt = 0x10,
+    Cobblestone = 0x11,
+    Ice = 0x12,
+    LeafyGrass = 0x13,
+    Salt = 0x14,
+    Limestone = 0x15,
+    Pavement = 0x16,
+}
+
+impl TryFrom<u8> for Material {
+    type Error = InvalidMaterialError;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        use Material::*;
+
+        Ok(match value {
+            0x00 => Air,
+            0x01 => Water,
+            0x02 => Grass,
+            0x03 => Slate,
+            0x04 => Concrete,
+            0x05 => Brick,
+            0x06 => Sand,
+            0x07 => WoodPlanks,
+            0x08 => Rock,
+            0x09 => Glacier,
+            0x0A => Snow,
+            0x0B => Sandstone,
+            0x0C => Mud,
+            0x0D => Basalt,
+            0x0E => Ground,
+            0x0F => CrackedLava,
+            0x10 => Asphalt,
+            0x11 => Cobblestone,
+            0x12 => Ice,
+            0x13 => LeafyGrass,
+            0x14 => Salt,
+            0x15 => Limestone,
+            0x16 => Pavement,
+            _ => return Err(InvalidMaterialError(value)),
+        })
+    }
+}
+
+#[derive(Debug)]
+pub struct InvalidMaterialError(u8);
+
+impl std::error::Error for InvalidMaterialError {}
+
+impl fmt::Display for InvalidMaterialError {
+    fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+        write!(formatter, "Invalid binary material value {:x?}", self.0)
+    }
+}
```

Comment:
We have `thiserror` for this. You should be able to `derive(Error)` for `InvalidMaterialError`.

Really though we want to use the `Error` type from the `error` module in this crate. You can go look at `MaterialColors` or `Attributes` for an idea of how this works.

Reply:
i copypasted this from InvalidTypeError in rbx_binary iirc i dont know why i didnt use the error module for this